File: crates/card/rank.rs
use enum_trait::Enum;

use crate::*;

#[derive(Clone, Copy, Debug, Enum, Eq, Ord, PartialEq, PartialOrd)]
pub enum Rank {
    Ace,
    Two,
    Three,
    Four,
    Five,
    Six,
    Seven,
    Eight,
    Nine,
    Ten,
    Jack,
    Queen,
    King,
}

pub use Rank::*;

impl Rank {
    pub fn to_char(self) -> char {
        match self {
            Ace => 'A',
            Two => '2',
            Three => '3',
            Four => '4',
            Five => '5',
            Six => '6',
            Seven => '7',
            Eight => '8',
            Nine => '9',
            Ten => 'T',
            Jack => 'J',
            Queen => 'Q',
            King => 'K',
        }
    }

    pub fn to_str(self) -> &'static str {
        match self {
            Ace => "Ace",
            Two => "Two",
            Three => "Three",
            Four => "Four",
            Five => "Five",
            Six => "Six",
            Seven => "Seven",
            Eight => "Eight",
            Nine => "Nine",
            Ten => "Ten",
            Jack => "Jack",
            Queen => "Queen",
            King => "King",
        }
    }

    pub fn with_suit(self, suit: Suit) -> Card {
        Card::new(self, suit)
    }
}

impl std::convert::From<Card> for Rank {
    fn from(card: Card) -> Self {
        card.rank()
    }
}

impl std::convert::TryFrom<char> for Rank {
    type Error = ParseError;

    fn try_from(char: char) -> Result<Self, Self::Error> {
        match char {
            'A' => Ok(Ace),
            '2' => Ok(Two),
            '3' => Ok(Three),
            '4' => Ok(Four),
            '5' => Ok(Five),
            '6' => Ok(Six),
            '7' => Ok(Seven),
            '8' => Ok(Eight),
            '9' => Ok(Nine),
            'T' => Ok(Ten),
            'J' => Ok(Jack),
            'Q' => Ok(Queen),
            'K' => Ok(King),
            _ => Err(ParseError::Invalid),
        }
    }
}

impl std::convert::From<Rank> for char {
    fn from(rank: Rank) -> Self {
        rank.to_char()
    }
}

impl std::convert::From<Rank> for &'static str {
    fn from(rank: Rank) -> Self {
        rank.to_str()
    }
}

impl std::fmt::Display for Rank {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(formatter, "{}", self.to_str())
    }
}

impl std::str::FromStr for Rank {
    type Err = ParseError;

    fn from_str(str: &str) -> Result<Self, Self::Err> {
        parse_char(str).and_then(Rank::try_from)
    }
}

File: packages/app/sources/models/Rank.ts
export enum Rank {
	Ace = 0,
	Two = 1,
	Three = 2,
	Four = 3,
	Five = 4,
	Six = 5,
	Seven = 6,
	Eight = 7,
	Nine = 8,
	Ten = 9,
	Jack = 10,
	Queen = 11,
	King = 12,
}

export namespace Rank {
	export function toCharacter(rank: Rank): string {
		switch (rank) {
			case Rank.Ace:
				return "A";
			case Rank.Two:
				return "2";
			case Rank.Three:
				return "3";
			case Rank.Four:
				return "4";
			case Rank.Five:
				return "5";
			case Rank.Six:
				return "6";
			case Rank.Seven:
				return "7";
			case Rank.Eight:
				return "8";
			case Rank.Nine:
				return "9";
			case Rank.Ten:
				return "T";
			case Rank.Jack:
				return "J";
			case Rank.Queen:
				return "Q";
			case Rank.King:
				return "K";
		}
	}

	// rome-ignore lint/suspicious/noShadowRestrictedNames: Not a problem in a namespace.
	export function toString(rank: Rank): string {
		switch (rank) {
			case Rank.Ace:
				return "Ace";
			case Rank.Two:
				return "Two";
			case Rank.Three:
				return "Three";
			case Rank.Four:
				return "Four";
			case Rank.Five:
				return "Five";
			case Rank.Six:
				return "Six";
			case Rank.Seven:
				return "Seven";
			case Rank.Eight:
				return "Eight";
			case Rank.Nine:
				return "Nine";
			case Rank.Ten:
				return "Ten";
			case Rank.Jack:
				return "Jack";
			case Rank.Queen:
				return "Queen";
			case Rank.King:
				return "King";
		}
	}
}
